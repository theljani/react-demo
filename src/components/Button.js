import React from 'react';
import PropTypes from 'prop-types';

const Button = ({label, btnClasses, onClick}) => {
  return(
    <button className={`btn ${btnClasses}`} onClick={onClick}>{label}</button>
  )
};

Button.proptypes = {
  label: PropTypes.string.isRequired,
  btnClasses: PropTypes.string,
  onClick: PropTypes.func.isRequired
};

export default Button;