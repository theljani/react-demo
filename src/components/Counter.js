


import React from 'react';
import { connect } from 'react-redux';

import Input from './InputField';
import Button from './Button';

class Counter extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      counter: 0,
      errors: []
    };

    this.onChange = this.onChange.bind(this);
    this.onIncrement = this.onIncrement.bind(this);
    this.onDecrement = this.onDecrement.bind(this);
    this.onCounterBlur = this.onCounterBlur.bind(this);
  }

  onChange = (event) => {
    event.persist();
    this.setState(prevState => ({
      [event.target.name]: event.target.value
    }));
  }

  onIncrement = (event) => {
    event.preventDefault();
    this.setState(prevState => ({
      counter: parseInt(prevState.counter) + 1
    }));
  }

  onDecrement = (event) => {
    event.preventDefault();
    this.setState(prevState => ({
      counter: parseInt(prevState.counter) - 1
    }));
  }

  onCounterBlur = (event) => {
    event.persist();
    if(!event.target.value) {
      debugger
      this.setState(prevState =>({
        errors: ['Counter value is mandatory']
      }));
    }
  }


  render() {
    return (
      <div className="container">
        <div className="form-group">
          <div className="col-md-12">
            <div className='row'>
              <div className="offset-md-2 col-md-8">
                <Input
                  name="counter"
                  label="Counter"
                  type="text"
                  value={this.state.counter}
                  onChange={this.onChange}
                  onFocus ={() => {}}
                  onBlur={this.onCounterBlur}
                  errors={this.state.errors} />
                <br />
                <p>Current counter is: {this.state.counter}</p>
              </div>
              <div className="offset-md-2 col-md-8">
                <Button label="Increment" btnClasses="btn-primary" onClick={this.onIncrement} />
                {' '}
                <Button label="Decrement" btnClasses="btn-warning" onClick={this.onDecrement} />
              </div>
            </div>
          </div>
        </div>
      </div>
    )
  }

}

export default Counter;