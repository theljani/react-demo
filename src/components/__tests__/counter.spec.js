jest.unmock('../Counter');
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';

// Components
import Counter from '../Counter';
import InputFied from '../InputField';
import Button from '../Button';

describe('The counter compponent', () => {

  it('Should render properly', () => {
    
    const wrapper = shallow(<Counter />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should initialise state properly', () => {
    const wrapper = shallow(<Counter />);

    expect(wrapper.state().counter).toEqual(0);
  });

  it('Should increment by 1 when invoking onIncrement method', () => {
    const wrapper = shallow(<Counter />);
    const event = {
      preventDefault: () => {}
    };

    const instance = wrapper.instance();

    instance.onIncrement(event);
    expect(wrapper.state().counter).toBe(1);
  })

  it('Should increment by 1 when invoking onDecrement method', () => {
    const wrapper = shallow(<Counter />);
    const event = {
      preventDefault: () => {}
    };
    const instance = wrapper.instance();

    instance.onIncrement(event);
    instance.onIncrement(event);
    instance.onDecrement(event);
    expect(wrapper.state().counter).toBe(1);
  })

  it('Should contains an InputField component as a child', () => {
    const wrapper = mount(<Counter />); 
    const inputField = wrapper.find(InputFied);

    expect(inputField).toBeDefined();
    expect(inputField).toHaveLength(1);
  });

  it('Should contains an InputField component as a child and progarate props properly', () => {
    const wrapper = mount(<Counter />); 
    const inputField = wrapper.find(InputFied);
    expect(inputField).toBeDefined();
    expect(inputField.props()).toBeDefined();
    expect(inputField.props().value).toEqual(0);
    expect(inputField.props().value).toEqual(0);
    expect(inputField.props().name).toEqual('counter');
    expect(inputField.props().label).toEqual('Counter');
    expect(inputField.props().errors).toBeDefined();
    expect(inputField.props().errors).toHaveLength(0);
    expect(inputField.props().onChange).toBeDefined();
    expect(inputField.props().onBlur).toBeDefined();
    expect(inputField.props().onFocus).toBeDefined();
  });

  it('Should contains 2 Button components as children', () => {
    const counterComponent = mount(<Counter />);

    expect(counterComponent).toBeDefined();
    expect(counterComponent).toMatchSnapshot();

    const buttonComponents = counterComponent.find(<Button />);

    expect(buttonComponents).toBeDefined();

    console.log(buttonComponents.first());
  });
});

