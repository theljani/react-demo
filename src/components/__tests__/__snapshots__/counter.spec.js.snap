// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The counter compponent Should contains 2 Button components as children 1`] = `
<Counter>
  <div
    className="container"
  >
    <div
      className="form-group"
    >
      <div
        className="col-md-12"
      >
        <div
          className="row"
        >
          <div
            className="offset-md-2 col-md-8"
          >
            <Input
              errors={Array []}
              label="Counter"
              name="counter"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="text"
              value={0}
            >
              <div
                className="form-group "
              >
                <label
                  className="form-control-label"
                  htmlFor="counter"
                >
                  Counter
                </label>
                <input
                  className="form-control "
                  name="counter"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  type="text"
                  value={0}
                />
              </div>
            </Input>
            <br />
            <p>
              Current counter is: 
              0
            </p>
          </div>
          <div
            className="offset-md-2 col-md-8"
          >
            <Button
              btnClasses="btn-primary"
              label="Increment"
              onClick={[Function]}
            >
              <button
                className="btn btn-primary"
                onClick={[Function]}
              >
                Increment
              </button>
            </Button>
             
            <Button
              btnClasses="btn-warning"
              label="Decrement"
              onClick={[Function]}
            >
              <button
                className="btn btn-warning"
                onClick={[Function]}
              >
                Decrement
              </button>
            </Button>
          </div>
        </div>
      </div>
    </div>
  </div>
</Counter>
`;

exports[`The counter compponent Should render properly 1`] = `
<div
  className="container"
>
  <div
    className="form-group"
  >
    <div
      className="col-md-12"
    >
      <div
        className="row"
      >
        <div
          className="offset-md-2 col-md-8"
        >
          <Input
            errors={Array []}
            label="Counter"
            name="counter"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value={0}
          />
          <br />
          <p>
            Current counter is: 
            0
          </p>
        </div>
        <div
          className="offset-md-2 col-md-8"
        >
          <Button
            btnClasses="btn-primary"
            label="Increment"
            onClick={[Function]}
          />
           
          <Button
            btnClasses="btn-warning"
            label="Decrement"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
